### PR TITLE
Refactoring: Backport class name generation improvements from autospidergen branch

### PR DESCRIPTION
**Pull request by CloCkWeRX**
Cherry pick 5e5a936785581a7dc8447db9bfdd817522483483 for class name generation

Another small chunk of https://github.com/alltheplaces/alltheplaces/pull/6526

### CODE REVIEW

**Review by davidhicks — APPROVED** [review]

File: locations/name_suggestion_index.py [hunk 1]
```diff
@@ -35,7 +42,37 @@ def _ensure_loaded(self):
...
+                official_website_domain = urlparse(official_website).netloc
+                if official_website_domain == supplied_url_domain:
+                    return wikidata_code
+        # Next attempt to find an exact match excluding any "www." prefix
+        for wikidata_code, org_parameters in self.wikidata_json.items():
+            for official_website in org_parameters.get("officialWebsites", []):
+                official_website_domain = urlparse(official_website).netloc
+                if official_website_domain.lstrip("www.") == supplied_url_domain.lstrip("www."):
+                    return wikidata_code
+        # Last attempt to find a fuzzy match for registered domain (exlcuding subdomains)
```
Comment (CloCkWeRX) [inline locations/name_suggestion_index.py:66]:
```suggestion
        # Last attempt to find a fuzzy match for registered domain (excluding subdomains)
```

File: tests/test_name_suggestion_index.py [hunk 2]
```diff
@@ -27,3 +27,26 @@ def test_iter_nsi():
...
     assert i["tags"]["amenity"] == "fast_food"
+
+
+def test_generate_keys_from_nsi_attributes():
+    nsi = NSI()
+    matches = list(nsi.iter_nsi("Q38076"))
+    keys_to_find = [
+        ("mcdonalds", "McDonaldsSpider"),
+        ("mcdonalds_fr", "McDonaldsFRSpider"),
+        ("mai_dang_lao_mcdonalds_hk_mo", "MaiDangLaoMcDonaldsHKMOSpider"),
```
Comment (davidhicks) [inline tests/test_name_suggestion_index.py:38]: `MaiDangLaoMcDonaldsHKMOSpider` should now be `MaiDangLaoMcdonaldsHKMOSpider`... I think... after the recent changes of #8829 to automatically fix/set spider class names based on the spider `name` field.

**Review by davidhicks — CHANGES_REQUESTED** [review]

File: tests/test_name_suggestion_index.py [hunk 3]
```diff
@@ -27,3 +27,26 @@ def test_iter_nsi():
...
     assert i["displayName"] == "Greggs"
     assert i["tags"]["amenity"] == "fast_food"
+
+
+def test_generate_keys_from_nsi_attributes():
+    nsi = NSI()
+    matches = list(nsi.iter_nsi("Q38076"))
+    keys_to_find = [
+        ("mcdonalds", "McDonaldsSpider"),
+        ("mcdonalds_fr", "McDonaldsFRSpider"),
```
Comment (CloCkWeRX) [inline tests/test_name_suggestion_index.py:37]:
```suggestion
        ("mcdonalds_fr", "McdonaldsFRSpider"),
```

File: tests/test_name_suggestion_index.py [hunk 4]
```diff
@@ -27,3 +27,26 @@ def test_iter_nsi():
     i = matches[0]
     assert i["displayName"] == "Greggs"
     assert i["tags"]["amenity"] == "fast_food"
+
+
+def test_generate_keys_from_nsi_attributes():
+    nsi = NSI()
+    matches = list(nsi.iter_nsi("Q38076"))
+    keys_to_find = [
+        ("mcdonalds", "McDonaldsSpider"),
```
Comment (CloCkWeRX) [inline tests/test_name_suggestion_index.py:36]:
```suggestion
        ("mcdonalds", "McdonaldsSpider"),
```

File: tests/test_name_suggestion_index.py [hunk 5]
```diff
@@ -27,3 +27,26 @@ def test_iter_nsi():
...
     assert i["tags"]["amenity"] == "fast_food"
+
+
+def test_generate_keys_from_nsi_attributes():
+    nsi = NSI()
+    matches = list(nsi.iter_nsi("Q38076"))
+    keys_to_find = [
+        ("mcdonalds", "McdonaldsSpider"),
+        ("mcdonalds_fr", "McdonaldsFRSpider"),
+        ("mai_dang_lao_mcdonalds_hk_mo", "MaiDangLaoMcDonaldsHKMOSpider"),
```
Comment (CloCkWeRX) [inline tests/test_name_suggestion_index.py:38]:
```suggestion
        ("mai_dang_lao_mcdonalds_hk_mo", "MaiDangLaoMcdonaldsHKMOSpider"),
```

File: tests/test_name_suggestion_index.py [hunk 6]
```diff
@@ -27,3 +27,26 @@ def test_iter_nsi():
     i = matches[0]
     assert i["displayName"] == "Greggs"
     assert i["tags"]["amenity"] == "fast_food"
+
+
+def test_generate_keys_from_nsi_attributes():
+    nsi = NSI()
+    matches = list(nsi.iter_nsi("Q38076"))
+    keys_to_find = [
+        ("mcdonalds", "McdonaldsSpider"),
```
Comment (CloCkWeRX) [inline tests/test_name_suggestion_index.py:36]:
```suggestion
        # NOTE: Spider naming convention tools ignore english naming conventions. While there are some inconsistencies in what humans would choose to do, that is acceptable.
        ("mcdonalds", "McdonaldsSpider"),
```

**Comment by CloCkWeRX** [comment]
Thinking about this one, do we just want to fix the spider naming consistency check to be smarter for:

- Acronyms (BNPFooSpider)
- Common name capitalisation patterns (McDonalds)
- Country codes at end of spider (Brand It)

Or do we want to have our code smush the human naming from the NSI into a class name?

From the ruby world; the former would be done by:

- Add and maintain a list of text inflection capitalisatons - https://stackoverflow.com/questions/803262/how-do-i-tell-rails-how-to-correctly-capitalize-an-acronym

https://github.com/ixmatus/inflector/blob/master/tests/tests.py looks like a python port

**Pull request closed without merging**